pypeit/datamodel.py

```diff
@@ -0,0 +1,942 @@
+
+        data = DubiousInitContainer(x,y)
+        print(np.array_equal(data.out, data.inp1+data.inp2))
+        # True
+
+    - One component of the data model wasn't instantiated, so it will
+      be None::
+        
+        print(data.alt is None)
+        # True
+
+    - The problem with ``DubiousInitContainer`` is that it has
+      attributes that cannot be reinstantiated from what's written to
+      disk. That is, :class:`DataContainers` aren't really fully
+      formed objects unless all of its relevant attributes are
+      components of the data model::
+
+        _data = DubiousInitContainer.from_hdu(hdu)
+        print(_data.func == DubiousInitContainer(x,y).func)
+        # False
+    
+    - This is solved by adding func to the datamodel::
+    
+        data = ComplexInitContainer(x,y)
+        _data = ComplexInitContainer.from_hdu(data.to_hdu(add_primary=True))
+        print(data.func == _data.func)
+        # True
+
+----
+
+.. include common links, assuming primary doc root is up one directory
+.. include:: ../links.rst
+"""
+import os
+import warnings
+
+from IPython import embed
+
+import numpy as np
+
+from astropy.io import fits
+from astropy.table import Table
+
+from pypeit import io
+
+class DataContainer:
+    """
+    Defines an abstract class for holding and manipulating data.
+
+    The primary utilities of the class are:
+        - Attributes can be accessed normally or as expected for a :obj:`dict`
+        - Attributes and items are restricted to conform to a specified data model.
+
+    This abstract class should only be used as a base class.
+
+    Derived classes must do the following:
+
+        - Define a datamodel
+        - Provide an :func:`__init__` method that defines the
+          instantiation calling sequence and passes the relavant
+          dictionary to this base-class instantiation.
+        - Provide a :func:`_validate` method, if necessary, that
+          processes the data provided in the `__init__` into a
+          complete instantiation of the object. This method and the
+          :func:`__init__` method are the *only* places where
+          attributes can be added to the class.
+        - Provide a :func:`_bundle` method that reorganizes the
+          datamodel into partitions that can be written to one or
+          more fits extensions. More details are provided in the
+          description of :func:`_bundle`.
+        - Provide a :func:`_parse` method that parses information in
+          one or more fits extensions into the appropriate datamodel.
+          More details are provided in the description of
+          :func:`_parse`.
+
+    Args:
+        d (:obj:`dict`):
+            Dictionary to copy to the internal attribute dictionary.
+    """
+
+    # Define the data model
+    datamodel = None
+    """
+    Provides the class data model. This is undefined in the abstract
+    class and should be overwritten in the derived classes.
+
```

Comment:
While it may be a little repetitive, you might describe the standard format for `datamodel`, i.e.  a `dict` with keys `otype`, `descr` and optionally `atype`, etc.

Reply:
Done.  Copied it from the module docstring.